src/transformers/generation/utils.py

```diff
@@ -1511,7 +1518,12 @@ def generate(
         accepts_attention_mask = "attention_mask" in set(inspect.signature(self.forward).parameters.keys())
         requires_attention_mask = "encoder_outputs" not in model_kwargs
         kwargs_has_attention_mask = model_kwargs.get("attention_mask", None) is not None
-        self._prepare_special_tokens(generation_config, kwargs_has_attention_mask)
+
+        device = None
+        if "input_ids" in model_kwargs and isinstance(model_kwargs["input_ids"], torch.Tensor):
+            device = model_kwargs["input_ids"].device
+
+        self._prepare_special_tokens(generation_config, kwargs_has_attention_mask, device=device)
```

Comment:
If I get it right: the `device` comes from the main model input, and not from the model itself. 

Assuming what I wrote above is correct, we should get the `device` variable after the `_prepare_model_inputs` call, which extracts the main model input from the different keywords we might see (for instance, `Whisper` does not use `input_ids`). In that case, I would move these lines to after L1532 (currently `batch_size = inputs_tensor.shape[0]`), and use `device=inputs_tensor.device` :D

Reply:
Makes totally sense! Done!